presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -230,90 +217,97 @@ private static Partitioner createPartitioner(PartitioningHandle partitioning, in
+            PartitioningHandle partitioning = fragment.getPartitioning();
+            taskSourceRdd = Optional.of(createTaskSourcesRdd(sparkContext, session, partitioning, tableScans, numberOfInputPartitions).rdd());
+        }
+        else if (rddInputs.size() == 0) {
+            checkArgument(fragment.getPartitioning().equals(SINGLE_DISTRIBUTION), "SINGLE_DISTRIBUTION partitioning is expected: %s", fragment.getPartitioning());
+            taskSourceRdd = Optional.of(new PrestoSparkTaskSourceRdd(sparkContext.sc(), ImmutableList.of(ImmutableList.of())));
+        }
+        else {
+            taskSourceRdd = Optional.empty();
+        }
+
         return JavaPairRDD.fromRDD(
-                new PrestoSparkZipRdd(sparkContext.sc(), rdds.build(), taskProcessor),
+                new PrestoSparkTaskRdd(sparkContext.sc(), taskSourceRdd, inputRdds.build(), taskProcessor),
                 classTag(Integer.class),
                 classTag(PrestoSparkRow.class));
     }
 
-    private JavaPairRDD<Integer, PrestoSparkRow> createSourceRdd(
+    private JavaRDD<SerializedTaskSource> createTaskSourcesRdd(
             JavaSparkContext sparkContext,
             Session session,
-            PlanFragment fragment,
-            PrestoSparkTaskExecutorFactoryProvider executorFactoryProvider,
-            CollectionAccumulator<SerializedTaskStats> taskStatsCollector,
-            TableWriteInfo tableWriteInfo,
-            Map<PlanFragmentId, Broadcast<List<PrestoSparkSerializedPage>>> broadcastInputs)
+            PartitioningHandle partitioning,
+            List<TableScanNode> tableScans,
+            Optional<Integer> expectedNumberOfPartitions)
     {
-        checkInputs(fragment.getRemoteSourceNodes(), ImmutableMap.of(), broadcastInputs);
-
-        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
-        checkArgument(
-                tableScans.size() == 1,
-                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
-                fragment.getId(),
-                tableScans.size());
-
-        TableScanNode tableScan = getOnlyElement(tableScans);
-
-        List<ScheduledSplit> splits = getSplits(session, tableScan);
-        shuffle(splits);
-        int initialPartitionCount = getSparkInitialPartitionCount(session);
-        int numTasks = Math.min(splits.size(), initialPartitionCount);
-        if (numTasks == 0) {
-            return JavaPairRDD.fromJavaRDD(sparkContext.emptyRDD());
+        ListMultimap<Integer, TaskSource> taskSourcesMap = ArrayListMultimap.create();
+        for (TableScanNode tableScan : tableScans) {
+            List<ScheduledSplit> scheduledSplits = getSplits(session, tableScan);
+            shuffle(scheduledSplits);
+            SetMultimap<Integer, ScheduledSplit> assignedSplits = assignSplitsToTasks(session, partitioning, scheduledSplits);
+            asMap(assignedSplits).forEach((partitionId, splits) ->
```

Comment:
nit: why not just using a `for-each` loop? ;)

Reply:
I like this style as you can give a key and a value a meaningful name

> (partitionId, splits) ->